### PR TITLE
Type-checker errors interface

### PR DESCRIPTION
**Pull request by anatoliykmetyuk**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/Inliner.scala [hunk 1]
```diff
@@ -67,6 +67,7 @@ object Inliner {
    */
   def inlineCall(tree: Tree)(implicit ctx: Context): Tree = {
     if (tree.symbol == defn.CompiletimeTesting_typeChecks) return Intrinsics.typeChecks(tree)
+    if (tree.symbol == defn.CompiletimeTesting_typeCheckErrors) return Intrinsics.typeCheckErrors(tree)
```
Comment (anatoliykmetyuk) [inline compiler/src/dotty/tools/dotc/typer/Inliner.scala]: Check that `tree.symbol.owner` is the `compiletime.testing` package.

File: compiler/src/dotty/tools/dotc/typer/Inliner.scala [hunk 2]
```diff
@@ -205,20 +208,51 @@ object Inliner {
...
+      constructor.appliedTo(
+        lit(error.message),
+        lit(error.pos.lineContent.reverse.dropWhile("\n ".contains).reverse),
+        lit(error.pos.column),
+        if kind == ErrorKind.Parser then parserErrorKind else typerErrorKind)
+
+    private def packErrors(errors: List[(ErrorKind, Error)])(given Context): Tree =
+      val individualErrors: List[Tree] = errors.map(packError)
+      val errorTpt = ref(defn.CompiletimeTesting_ErrorClass)
+      ref(defn.ListModule).select(nme.apply)
```
Comment (anatoliykmetyuk) [inline compiler/src/dotty/tools/dotc/typer/Inliner.scala]: Create a helper with this logic in `tpd`. Replace `liftList` from `ReifyQuotes.scala` with it.

Comment (anatoliykmetyuk) [reply compiler/src/dotty/tools/dotc/typer/Inliner.scala]: Also, check if we use this logic in any other place.

File: library/src/scala/compiletime/testing/package.scala [hunk 3]
```diff
@@ -4,6 +4,15 @@ import scala.quoted._
 
 package object testing {
 
+  /** Represents a compile-time error */
```
Comment (anatoliykmetyuk) [inline library/src/scala/compiletime/testing/package.scala]: Specify in the scaladoc that this interface is unstable from version to version.

File: library/src/scala/compiletime/testing/package.scala [hunk 4]
```diff
@@ -4,6 +4,15 @@ import scala.quoted._
 
 package object testing {
```
Comment (anatoliykmetyuk) [inline library/src/scala/compiletime/testing/package.scala]: Remove this object, move everything into the outside.

File: compiler/src/dotty/tools/dotc/core/Definitions.scala [hunk 5]
```diff
@@ -389,6 +396,8 @@ class Definitions {
     methodNames.map(getWrapVarargsArrayModule.requiredMethod(_))
   })
 
+  @tu lazy val ListClass: ClassSymbol = ctx.requiredClass("scala.collection.immutable.List")
```
Comment (anatoliykmetyuk) [inline compiler/src/dotty/tools/dotc/core/Definitions.scala]: Can we avoid creating this one?

File: library/src/scala/compiletime/testing/Error.scala [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+/** Represents a compile-time error.
+ *
+ *  @see scala.compiletime.testing.typeCheckErrors
+ *
+ *  IMPORTANT: No stability guarantees are provided on the format of these
+ *  errors. This means the format and the API may change from
+ *  version to version. This API is to be used for testing purposes
+ *  only.
+ */
+case class Error(message: String, lineContent: String, column: Int, kind: ErrorKind)
```
Comment (bishabosha) [inline library/src/scala/compiletime/testing/Error.scala]:
```suggestion
final case class Error(message: String, lineContent: String, column: Int, kind: ErrorKind)
```

**Pull request merged**